src/Processors/Transforms/OptimizedPartialSortingTransform.cpp

```diff
@@ -0,0 +1,83 @@
+#include <Processors/Transforms/OptimizedPartialSortingTransform.h>
+#include <Interpreters/sortBlock.h>
+#include <Common/PODArray.h>
+
+namespace DB
+{
+
+OptimizedPartialSortingTransform::OptimizedPartialSortingTransform(
+    const Block & header_, SortDescription & description_, UInt64 limit_)
+    : ISimpleTransform(header_, header_, false)
+    , description(description_), limit(limit_)
+    , threshold_shared_block(nullptr)
+{
+}
+
+static ColumnRawPtrs extractColumns(const Block & block, const SortDescription& description)
+{
+    size_t size = description.size();
+    ColumnRawPtrs res;
+    res.reserve(size);
+
+    for (size_t i = 0; i < size; ++i)
+    {
+        const IColumn * column = !description[i].column_name.empty()
+            ? block.getByName(description[i].column_name).column.get()
+            : block.safeGetByPosition(description[i].column_number).column.get();
+        res.emplace_back(column);
+    }
+
+    return res;
+}
+
+void OptimizedPartialSortingTransform::transform(Chunk & chunk)
+{
+    if (read_rows)
+        read_rows->add(chunk.getNumRows());
+
+    auto block = getInputPort().getHeader().cloneWithColumns(chunk.detachColumns());
+    chunk.clear();
+
+    SharedBlockPtr shared_block = new detail::SharedBlock(std::move(block));
+    UInt64 rows_num = shared_block->rows();
+
+
+    if (threshold_shared_block) {
+        SharedBlockRowWithSortDescriptionRef row;
+        IColumn::Filter filter(rows_num);
```

Comment:
This memory is actually not zero. It is uninitialized as said failed test under tsan:
```
  Uninitialized value was created by a heap allocation
    #0 0x99dc20d in malloc (/usr/bin/clickhouse+0x99dc20d)
    #1 0x9bfaf5a in Allocator<false, false>::allocNoTrack(unsigned long, unsigned long) /build/obj-x86_64-linux-gnu/../src/Common/Allocator.h:221:27
    #2 0xd64ea2e in DB::PODArray<char8_t, 4096ul, Allocator<false, false>, 15ul, 16ul>::PODArray(unsigned long) (/usr/bin/clickhouse+0xd64ea2e)
    #3 0x24b752da in DB::OptimizedPartialSortingTransform::transform(DB::Chunk&) /build/obj-x86_64-linux-gnu/../src/Processors/Transforms/OptimizedPartialSortingTransform.cpp:47:25
```

Create empty `IColumn::Filter filter;` and call `filter.resize_fill(rows_num);` then.